pkg/capacity/capacity.go

```diff
@@ -382,6 +384,68 @@ func (c *Controller) onSCDelete(sc *storagev1.StorageClass) {
 	}
 }
 
+// refreshTopology identifies all work items matching the topology and schedules
+// a refresh. The node affinity is expected to come from controller.GenerateVolumeNodeAffinity,
+// i.e. only use NodeSelectorTerms and each of those must be based on the CSI driver's
+// topology key/value pairs of a topology segment.
+func (c *Controller) refreshTopology(nodeAffinity v1.VolumeNodeAffinity) {
+	if nodeAffinity.Required == nil || nodeAffinity.Required.NodeSelectorTerms == nil {
+		klog.Errorf("Capacity Controller: skipping refresh: unexpected VolumeNodeAffinity, missing NodeSelectorTerms: %v", nodeAffinity)
+		return
+	}
+
+	c.capacitiesLock.Lock()
+	defer c.capacitiesLock.Unlock()
+
+	for _, term := range nodeAffinity.Required.NodeSelectorTerms {
+		segment, err := termToSegment(term)
+		if err != nil {
+			klog.Errorf("Capacity Controller: skipping refresh: unexpected node selector term %+v: %v", term, err)
+			continue
+		}
+		for item := range c.capacities {
+			if item.segment.Compare(segment) == 0 {
+				klog.V(5).Infof("Capacity Controller: skipping refresh: enqueuing %+v because of the topology", item)
+				c.queue.Add(item)
+			}
+		}
+	}
+}
+
+func termToSegment(term v1.NodeSelectorTerm) (segment topology.Segment, err error) {
```

Comment:
Sounds good if there's another place where we need either logic in the future, we could consider a refactor at that time.